### PR TITLE
feat: implement onboarding

### PR DESCRIPTION
**Pull request by Victorsitou**
## Summary

https://github.com/discord/discord-api-docs/pull/5915

thanks to helish_88😈#7777 for helping me :)

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [ ] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: docs/api.rst
Comment (shiftinv) [inline docs/api.rst:3312]: This seems to include the `type` field as well

File: changelog/928.feature.rst
Comment (shiftinv) [inline changelog/928.feature.rst:7]:
```suggestion
```

File: docs/api.rst
Comment (shiftinv) [inline docs/api.rst:4881]: These aren't documented

File: docs/api.rst
Comment (shiftinv) [inline docs/api.rst:3361]:
```suggestion
        - :attr:`~AuditLogDiff.prompts`
        - :attr:`~AuditLogDiff.enabled`
```

File: docs/api.rst
Comment (shiftinv) [inline docs/api.rst:3372]: This appears to match the list for `onboarding_create`

File: docs/api.rst
Comment (shiftinv) [inline docs/api.rst:4861]:
```suggestion
        Whether the onboarding prompt option is part of the initial onboarding flow.
```

File: docs/api.rst
Comment (shiftinv) [inline docs/api.rst:4869]: Duplicate of `default_channels` below

File: docs/api.rst
Comment (shiftinv) [inline docs/api.rst:4893]:
```suggestion
    .. attribute:: prompts

        The list of onboarding prompts being changed.

        :type: List[:class:`OnboardingPrompt`]

    .. attribute:: default_channels

        The list of default channels for onboarding being changed.

        :type: List[:class:`~.abc.GuildChannel`]
```

File: tests/test_onboarding.py
Comment (shiftinv) [inline tests/test_onboarding.py:49]:
```suggestion
    onboarding_prompt_payload: onboarding_types.OnboardingPrompt = {
        "id": 0,
        "title": "test",
        "options": [],
        "single_select": True,
        "required": True,
        "in_onboarding": True,
        "type": OnboardingPromptType.multiple_choice.value,
    }

    return OnboardingPrompt._from_dict(
        data=onboarding_prompt_payload,
        state=mock.Mock(http=mock.AsyncMock()),
    )
```
This avoids the `type: ignore`s

File: disnake/enums.py
Comment (Victorsitou) [reply disnake/enums.py:390]: > These aren't documented, not sure if we want to implement them yet

I should definitely remove that if it's not documented, but I just didn't want to remove it in case it is documented.

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py:173]: Sounds good to me 👍

File: disnake/emoji.py
Comment (Victorsitou) [reply disnake/emoji.py]: a7846ef4ba2723894c0430aedd3ab6b2f932c3b5

File: disnake/http.py
Comment (Victorsitou) [reply disnake/http.py]: 46597446c818ca1d19dde27c0e6f1a915cbd5674

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py]: Alright! a62cd2ae6bd6a096437c5ea3914be7aca4f5eef9

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py]: 7bce38c395bb9884c70d794227611bf4b5e4c1d5

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py]: 90d9813fd1c211a08bee1a146003f24a90d0304f

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py]: cf92da34c7db1633ab9ed802f0c893db56654599

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py]: 2c704bd2874b1aa7be74ee7f31c42e137554296c

File: tests/test_onboarding.py
Comment (Victorsitou) [reply tests/test_onboarding.py]: thanks 4b0f2b3d90700a68f681572abd5b99132c4f93f0

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py]: Changed it in 1d8cd920ec3715f6a3c5b1eddb9e58420f764ea3, is it ok like that?

Comment (shiftinv) [reply disnake/onboarding.py]: With #980 this will be a lot simpler, should just be a matter of calling `guild._state._get_emoji_from_data(data["emoji"])` then.

File: docs/api.rst
Comment (Victorsitou) [reply docs/api.rst]: closing since audit logs were not added in api-docs pr. 2b81f4c90af51506ae7f6c0843f514b58be80f7a

File: disnake/enums.py
Comment (Victorsitou) [reply disnake/enums.py]: closing since audit logs were not added in api-docs pr. 2b81f4c90af51506ae7f6c0843f514b58be80f7a

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py]: Removed in d71866dcda85604dd5f00bb7df185381430746cb

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py]: Okay, I'll mark this as resolved then

File: disnake/onboarding.py
Comment (shiftinv) [inline disnake/onboarding.py:226]: `_from_dict` here and in `OnboardingPromptOption` can be merged into `__init__`, now that these don't have to be user-constructible

File: disnake/onboarding.py
Comment (shiftinv) [inline disnake/onboarding.py:68]: Instead of the guild ID, this should include the guild itself.
Additionally, `default_channel_ids` is a bit too verbose to include here.

File: disnake/onboarding.py
Comment (shiftinv) [inline disnake/onboarding.py:115]: see other comment about `default_channel_ids`, I feel like something similar could be done here

File: disnake/onboarding.py
Comment (shiftinv) [inline disnake/onboarding.py:134]: Don't think this should contain `roles` and `channels`, they're also rather verbose

File: disnake/types/onboarding.py
Comment (shiftinv) [inline disnake/types/onboarding.py:14]: `description` is nullable, according to the docs - I haven't seen that happen, but it should be handled regardless

File: docs/api.rst
Comment (shiftinv) [inline docs/api.rst:6059]:
```suggestion
.. autoclass:: ApplicationRoleConnectionMetadata
    :members:
```
looks like this one disappeared when the other two got added

File: disnake/onboarding.py
Comment (shiftinv) [inline disnake/onboarding.py:147]:
```suggestion
        if emoji_data := data.get("emoji"):
            emoji = guild._state.get_reaction_emoji(emoji_data)
        else:
            emoji = None
```
This is a bit simpler - even if the name doesn't really make sense before #980 is merged, it's fine for now :p

File: disnake/onboarding.py
Comment (shiftinv) [reply disnake/onboarding.py]: followed up in https://github.com/DisnakeDev/disnake/pull/928#discussion_r1155160553

File: changelog/928.feature.rst
Comment (Victorsitou) [reply changelog/928.feature.rst:4]: ah yes I forgot

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py:226]: you're right, my bad.

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py]: dae932be51d4522f81a40a11a416f251f3bc2354

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py]: dae932be51d4522f81a40a11a416f251f3bc2354

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py]: dae932be51d4522f81a40a11a416f251f3bc2354

File: disnake/types/onboarding.py
Comment (Victorsitou) [reply disnake/types/onboarding.py]: 07d839b85a81f1facceaf2323833b3ab7af85ec7

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py:151]: 52a487cb3c3d3f30bd38465183e6cb22e53828bb

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py]: dae932be51d4522f81a40a11a416f251f3bc2354

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py]: 7466f3a7aeff7868b8ef8c84e078ca0fd523cb60

File: docs/api.rst
Comment (shiftinv) [inline docs/api.rst:6077]: these two should be put under Discord Models, I think

File: disnake/onboarding.py
Comment (shiftinv) [inline disnake/onboarding.py:121]: could you add type annotations to these, as well as the ones in the other two classes?

**Review by shiftinv — COMMENTED** [review]
other than the occasional ci errors and these last two comments, lgtm!
I haven't tested everything yet, will approve once that's done

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py:121]: Alright!

File: docs/api.rst
Comment (Victorsitou) [reply docs/api.rst]: c2ce2cc76acec06a265cfdc38e811315aeceb856

File: disnake/onboarding.py
Comment (Victorsitou) [reply disnake/onboarding.py]: d9d0c00772c29d379e72a40b9711581a36588b8b

File: disnake/onboarding.py
Comment (shiftinv) [inline disnake/onboarding.py:144]:
```suggestion
    emoji: Optional[Union[:class:`PartialEmoji`, :class:`Emoji`, :class:`str`]]
```
(`get_reaction_emoji` typings are slightly incorrect, fixed in #980)

File: disnake/onboarding.py
Comment (shiftinv) [inline disnake/onboarding.py:178]:
```suggestion
        self.emoji: Optional[Union[Emoji, PartialEmoji, str]]
```
see first comment

File: disnake/onboarding.py
Comment (shiftinv) [inline disnake/onboarding.py:163]:
```suggestion
```

**Review by shiftinv — APPROVED** [review]
thanks!

File: changelog/928.feature.rst
Comment (shiftinv) [inline changelog/928.feature.rst:3]:
```suggestion
    - Add :meth:`Guild.onboarding`.
```

**Pull request merged**
